aea/protocols/dialogue/base.py

```diff
@@ -30,13 +30,65 @@
 from abc import ABC
 from enum import Enum
 from inspect import signature
-from typing import Callable, Dict, FrozenSet, List, Optional, Set, Tuple, Type, cast
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    FrozenSet,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    cast,
+)
 
 from aea.common import Address
-from aea.exceptions import enforce
+from aea.exceptions import AEAEnforceError, enforce
 from aea.protocols.base import Message
 
 
+class CompactMessage:
+    """Representation of a message in a dialogue."""
+
+    def __init__(
+        self,
+        performative: Message.Performative,
+        contents: Dict[str, Any],
+        is_incoming: Optional[bool] = None,
+        target: Optional[int] = None,
+    ) -> None:
+        """Initialise."""
+        self._performative = performative
+        self._contents = contents
+        self._is_incoming = is_incoming
+        self._target = target
+
+    @property
+    def is_incoming(self) -> bool:
+        """Get message's is_incoming."""
+        if self._is_incoming is None:
+            raise AEAEnforceError("is_incoming is None.")
+        return self._is_incoming
+
+    @property
+    def target(self) -> int:
+        """Get message's target."""
+        if self._target is None:
+            raise AEAEnforceError("target is None.")
+        return self._target
+
+    @property
+    def performative(self) -> Message.Performative:
+        """Get message's performative."""
+        return self._performative
+
+    @property
+    def contents(self) -> Dict[str, Any]:
```

Comment:
At a minimum this class also needs one class method `from_tuple` and one instance method `to_tuple`

Reply:
but the other question is whether we actually need to define a class or whether it is enough to define a type using typing or something like a namedtuple (https://docs.python.org/3/library/collections.html#collections.namedtuple)

Reply:
NamedTuple seems to be a good fit here.

Reply:
The only thing is we cannot do type-checking on the elements.

Reply:
https://stackoverflow.com/questions/56780005/how-to-validate-namedtuple-values ...but brings overhead as well.